contracts/axelar-gateway/src/test_auth.rs

```diff
@@ -0,0 +1,545 @@
+#![cfg(test)]
+extern crate std;
+
+use axelar_soroban_interfaces::types::{
+    ProofSignature, ProofSigner, WeightedSigner, WeightedSigners,
+};
+use soroban_sdk::{
+    testutils::{Address as _, BytesN as _},
+    Address, Bytes, BytesN, Env, Vec,
+};
+
+use axelar_soroban_std::testutils::assert_invocation;
+
+use crate::{
+    auth::{self, initialize_auth},
+    contract::{AxelarGateway, AxelarGatewayClient},
+    testutils::{
+        self, generate_proof, generate_random_payload_and_hash, generate_signers_set, initialize,
+        randint,
+    },
+};
+
+fn setup_env<'a>() -> (Env, Address, AxelarGatewayClient<'a>) {
+    let env = Env::default();
+    env.mock_all_auths();
+
+    let contract_id = env.register_contract(None, AxelarGateway);
+    let client = AxelarGatewayClient::new(&env, &contract_id);
+
+    (env, contract_id, client)
+}
+
+#[test]
+fn test_initialize() {
+    let (env, _, client) = setup_env();
+    let user = Address::generate(&env);
+
+    initialize(&env, &client, user, randint(0, 10), randint(1, 10));
+}
+
+#[test]
+#[should_panic(expected = "Already initialized")]
+fn fails_if_already_initialized() {
+    let (env, _, client) = setup_env();
+    let user_one = Address::generate(&env);
+    let user_two = Address::generate(&env);
+
+    initialize(&env, &client, user_one, randint(0, 10), randint(1, 10));
+
+    // second initialization should panic
+    initialize(&env, &client, user_two, randint(0, 10), randint(1, 10));
+}
+
+#[test]
+#[should_panic(expected = "Error(Contract, #6)")]
+fn fails_with_empty_signer_set() {
+    let (env, contract_id, _client) = setup_env();
+
+    // create an empty WeightedSigners vector
+    let empty_signer_set = Vec::<WeightedSigners>::new(&env);
+    let domain_separator: BytesN<32> = BytesN::random(&env);
+    let previous_signer_retention = randint(0, 10) as u64;
+    let minimum_rotation_delay = 0;
+    let initial_signers = empty_signer_set;
+
+    // call should panic because signer set is empty
+    env.as_contract(&contract_id, || {
+        initialize_auth(
+            env.clone(),
+            domain_separator,
+            minimum_rotation_delay,
+            previous_signer_retention,
+            initial_signers,
+        );
+    })
+
+    // assert!(res.is_err());
+}
+
+#[test]
+fn validate_proof() {
+    let (env, contract_id, client) = setup_env();
+    let user = Address::generate(&env);
+
+    let signers = initialize(&env, &client, user, randint(0, 10), randint(1, 10));
+
+    let msg_hash = generate_random_payload_and_hash(&env);
+    let proof = generate_proof(&env, msg_hash.clone(), signers);
+
+    // validate_proof shouldn't panic
+    env.as_contract(&contract_id, || {
+        let latest_signer_set = auth::validate_proof(&env, msg_hash, proof);
+        assert!(latest_signer_set);
+    });
+}
+
+#[test]
+#[should_panic(expected = "Error(Contract, #6)")]
```

Comment:
It would be great if we could add a comment for the error to improve code readability, as suggested below:
```suggestion
#[should_panic(expected = "Error(Contract, #6)")] // AuthError::InvalidSigners
```

Reply:
In the general case I'd agree, but #30 adds the proper error handling (and depends on this PR), so I think it's fine to leave it as is since it's going to be followed up by a proper solution.